File: DL-GMM/DL-GMM.php
<?php
/* https://www.decentlab.com/greenhouse-multi-monitor-for-lorawan */

abstract class DecentlabDecoder
{
    const PROTOCOL_VERSION = 2;

    protected $sensors;

    public function decode($payload = '')
    {
        $bytes = hex2bin($payload);
        $parts = [];

        $parts['version'] = ord($bytes[0]);
        if ($parts['version'] != self::PROTOCOL_VERSION) {
            $parts['error'] = sprintf("protocol version %u doesn't match v2", $parts['version']);
            return $parts;
        }

        $parts['device_id'] = unpack('n', substr($bytes, 1))[1];
        $flags = unpack('n', substr($bytes, 3))[1];

        /* decode payload */
        $k = 5;
        foreach ($this->sensors as $sensor) {
            if (($flags & 1) == 1) {
                $x = [];
                /* convert data to 16-bit integer array */
                for ($j = 0; $j < $sensor['length']; $j++) {
                    array_push($x, unpack('n', substr($bytes, $k))[1]);
                    $k += 2;
                }

                /* decode sensor values */
                foreach ($sensor['values'] as $value) {
                    if ($value['convert'] != NULL) {
                        $parts[$value['name'] . '_value'] = $value['convert']($x);
                        if ($value['unit'] != NULL) {
                            $parts[$value['name'] . '_unit'] = $value['unit'];
                        }
                    }
                }
            }
            $flags >>= 1;
        }

        return $parts;
    }
}

class DL_GMM_Decoder extends DecentlabDecoder {
    
    public function __construct()
    {
        $this->sensors = [
            [
                'length' => 7,
                'values' => [
                    [
                        'name' => 'photosynthetically_active_radiation',
                        'convert' => function ($x) { return ($x[0] - 32768) / 10; },
                        'unit' => 'µmol⋅m⁻²⋅s⁻¹',
                    ],
                    [
                        'name' => 'air_temperature',
                        'convert' => function ($x) { return ($x[1] - 32768) / 100; },
                        'unit' => '°C',
                    ],
                    [
                        'name' => 'air_humidity',
                        'convert' => function ($x) { return ($x[2] - 32768) / 10; },
                        'unit' => '%',
                    ],
                    [
                        'name' => 'co2_concentration',
                        'convert' => function ($x) { return ($x[3] - 32768) / 1; },
                        'unit' => 'ppm',
                    ],
                    [
                        'name' => 'atmospheric_pressure',
                        'convert' => function ($x) { return ($x[4] - 32768) / 100; },
                        'unit' => 'kPa',
                    ],
                    [
                        'name' => 'vapor_pressure_deficit',
                        'convert' => function ($x) { return ($x[5] - 32768) / 100; },
                        'unit' => 'kPa',
                    ],
                    [
                        'name' => 'dew_point',
                        'convert' => function ($x) { return ($x[6] - 32768) / 100; },
                        'unit' => '°C',
                    ],
                ],
            ],
            [
                'length' => 1,
                'values' => [
                    [
                        'name' => 'battery_voltage',
                        'convert' => function ($x) { return $x[0] / 1000; },
                        'unit' => 'V',
                    ],
                ],
            ],
        ];
    }
}


$decoder = new DL_GMM_Decoder();
$payloads = [
    '02532b00038726892081148297a57380cf81700bbc',
    '02528500020bbc',
];

foreach($payloads as $payload) {
    var_dump($decoder->decode($payload));
}

File: DL-GMM/DL-GMM.py
# -*- coding: utf-8 -*-

# https://www.decentlab.com/greenhouse-multi-monitor-for-lorawan

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import struct
from math import log, floor
from base64 import binascii

PROTOCOL_VERSION = 2

SENSORS = [
    {'length': 7,
     'values': [{'name': 'Photosynthetically active radiation',
                 'convert': lambda x: (x[0] - 32768) / 10,
                 'unit': 'µmol⋅m⁻²⋅s⁻¹'},
                {'name': 'Air temperature',
                 'convert': lambda x: (x[1] - 32768) / 100,
                 'unit': '°C'},
                {'name': 'Air humidity',
                 'convert': lambda x: (x[2] - 32768) / 10,
                 'unit': '%'},
                {'name': 'CO2 concentration',
                 'convert': lambda x: (x[3] - 32768) / 1,
                 'unit': 'ppm'},
                {'name': 'Atmospheric pressure',
                 'convert': lambda x: (x[4] - 32768) / 100,
                 'unit': 'kPa'},
                {'name': 'Vapor pressure deficit',
                 'convert': lambda x: (x[5] - 32768) / 100,
                 'unit': 'kPa'},
                {'name': 'Dew point',
                 'convert': lambda x: (x[6] - 32768) / 100,
                 'unit': '°C'}]},
    {'length': 1,
     'values': [{'name': 'Battery voltage',
                 'convert': lambda x: x[0] / 1000,
                 'unit': 'V'}]}
]


def decode(msg, hex=False):
    """msg: payload as one of hex string, list, or bytearray"""
    bytes_ = bytearray(binascii.a2b_hex(msg)
                       if hex
                       else msg)
    version = bytes_[0]
    if version != PROTOCOL_VERSION:
        raise ValueError("protocol version {} doesn't match v2".format(version))

    devid = struct.unpack('>H', bytes_[1:3])[0]
    bin_flags = bin(struct.unpack('>H', bytes_[3:5])[0])
    flags = bin_flags[2:].zfill(struct.calcsize('>H') * 8)[::-1]

    words = [struct.unpack('>H', bytes_[i:i + 2])[0]
             for i
             in range(5, len(bytes_), 2)]

    cur = 0
    result = {'Device ID': devid, 'Protocol version': version}
    for flag, sensor in zip(flags, SENSORS):
        if flag != '1':
            continue

        x = words[cur:cur + sensor["length"]]
        cur += sensor["length"]
        for value in sensor['values']:
            if 'convert' not in value:
                continue

            result[value['name']] = {'value': value['convert'](x),
                                     'unit': value.get('unit', None)}

    return result


if __name__ == '__main__':

    import pprint
    payloads = [
        b'02532b00038726892081148297a57380cf81700bbc',
        b'02528500020bbc',
    ]
    for pl in payloads:
        pprint.pprint(decode(pl, hex=True))
        print("")
